runner/src/indexer-logger/indexer-logger.ts

```diff
@@ -0,0 +1,77 @@
+import format from 'pg-format';
+import { wrapError } from '../utility';
+import PgClient from '../pg-client';
+import { type DatabaseConnectionParameters } from '../provisioner/provisioner';
+import { LogLevel } from '../stream-handler/stream-handler';
+import { trace } from '@opentelemetry/api';
+
+export interface LogEntry {
+  blockHeight: number
+  logTimestamp: Date
+  logType: string
+  logLevel: LogLevel
+  message: string
+}
+
+export enum LogType {
+  SYSTEM = 'system',
+  USER = 'user',
+}
+export default class IndexerLogger {
+  tracer = trace.getTracer('queryapi-runner-indexer-logger');
+
+  private readonly pgClient: PgClient;
+  private readonly schemaName: string;
+  private readonly logInsertQueryTemplate: string = 'INSERT INTO %I.__logs (block_height, date, timestamp, type, level, message) VALUES %L';
+  private readonly loggingLevel: number;
+
+  constructor (
+    functionName: string,
+    loggingLevel: number,
+    databaseConnectionParameters: DatabaseConnectionParameters,
+    pgClientInstance: PgClient | undefined = undefined
+  ) {
+    const pgClient = pgClientInstance ?? new PgClient({
+      user: databaseConnectionParameters.username,
+      password: databaseConnectionParameters.password,
+      host: process.env.PGHOST,
+      port: Number(databaseConnectionParameters.port),
+      database: databaseConnectionParameters.database,
+    });
+
+    this.pgClient = pgClient;
+    this.schemaName = functionName.replace(/[^a-zA-Z0-9]/g, '_');
+    this.loggingLevel = loggingLevel;
+  }
+
+  private shouldLog (logLevel: LogLevel): boolean {
+    return logLevel >= this.loggingLevel;
+  }
+
+  async writeLogs (
+    logEntries: LogEntry | LogEntry[],
```

Comment:
```suggestion
    logEntries: LogEntry[],
```
Sorry, I know @darunrs already commented on this, but I would just make this take an array only. That way you don't need conditional logic to work with either type. The caller is still free to a single log entry, but it would have to be an array, which I don't think is too inconvenient.

Reply:
I kinda like the suggestion. The reason I went with Type | Type[] for DMlHandler was actually because its user facing and Monaco types can be wonky. So, I wanted to be flexible with the user code. In this case, the callers of this function are us, so we can expect that we write code correctly.